File: PdfClown-NetStandard/CliSamples/org/pdfclown/samples/cli/BookmarksParsingSample.cs
using org.pdfclown.documents;
using org.pdfclown.documents.files;
using actions = org.pdfclown.documents.interaction.actions;
using org.pdfclown.documents.interaction.navigation.document;
using org.pdfclown.files;
using org.pdfclown.objects;

using System;

namespace org.pdfclown.samples.cli
{
  /**
    <summary>This sample demonstrates how to inspect the bookmarks of a PDF document.</summary>
  */
  public class BookmarksParsingSample
    : Sample
  {
    public override void Run(
      )
    {
      // 1. Opening the PDF file...
      string filePath = PromptFileChoice("Please select a PDF file");
      using(var file = new File(filePath))
      {
        Document document = file.Document;
  
        // 2. Get the bookmarks collection!
        Bookmarks bookmarks = document.Bookmarks;
        if(!bookmarks.Exists())
        {Console.WriteLine("\nNo bookmark available (Outline dictionary not found).");}
        else
        {
          Console.WriteLine("\nIterating through the bookmarks collection (please wait)...\n");
          // 3. Show the bookmarks!
          PrintBookmarks(bookmarks);
        }
      }
    }

    private void PrintBookmarks(
      Bookmarks bookmarks
      )
    {
      if(bookmarks == null)
        return;

      foreach(Bookmark bookmark in bookmarks)
      {
        // Show current bookmark!
        Console.WriteLine("Bookmark '" + bookmark.Title + "'");
        Console.Write("    Target: ");
        PdfObjectWrapper target = bookmark.Target;
        if(target is Destination)
        {PrintDestination((Destination)target);}
        else if(target is actions::Action)
        {PrintAction((actions::Action)target);}
        else if(target == null)
        {Console.WriteLine("[not available]");}
        else
        {Console.WriteLine("[unknown type: " + target.GetType().Name + "]");}

        // Show child bookmarks!
        PrintBookmarks(bookmark.Bookmarks);
      }
    }

    private void PrintAction(
      actions::Action action
      )
    {
      /*
        NOTE: Here we have to deal with reflection as a workaround
        to the lack of type covariance support in C# (so bad -- any better solution?).
      */
      Console.WriteLine("Action [" + action.GetType().Name + "] " + action.BaseObject);
      if(action.Is(typeof(actions::GoToDestination<>)))
      {
        if(action.Is(typeof(actions::GotoNonLocal<>)))
        {
          FileSpecification destinationFile = (FileSpecification)action.Get("DestinationFile");
          if(destinationFile != null)
          {Console.WriteLine("      Filename: " + destinationFile.Path);}

          if(action is actions::GoToEmbedded)
          {
            actions::GoToEmbedded.PathElement target = ((actions::GoToEmbedded)action).DestinationPath;
            Console.WriteLine("      EmbeddedFilename: " + target.EmbeddedFileName + " Relation: " + target.Relation);
          }
        }
        Console.Write("      ");
        PrintDestination((Destination)action.Get("Destination"));
      }
      else if(action is actions::GoToURI)
      {Console.WriteLine("      URI: " + ((actions::GoToURI)action).URI);}
    }

    private void PrintDestination(
      Destination destination
      )
    {
      Console.WriteLine(destination.GetType().Name + " " + destination.BaseObject);
      Console.Write("        Page ");
      object pageRef = destination.Page;
      if(pageRef is Page)
      {
        Page page = (Page)pageRef;
        Console.WriteLine(page.Number + " [ID: " + page.BaseObject + "]");
      }
      else
      {Console.WriteLine((int)pageRef+1);}

      object location = destination.Location;
      if(location != null)
      {Console.WriteLine("        Location {0}", location);}

      double? zoom = destination.Zoom;
      if(zoom.HasValue)
      {Console.WriteLine("        Zoom {0}", zoom.Value);}
    }
  }
}

File: java/pdfclown.samples.cli/src/org/pdfclown/samples/cli/BookmarksParsingSample.java
package org.pdfclown.samples.cli;

import org.pdfclown.documents.Document;
import org.pdfclown.documents.Page;
import org.pdfclown.documents.files.FileSpecification;
import org.pdfclown.documents.interaction.actions.Action;
import org.pdfclown.documents.interaction.actions.GoToDestination;
import org.pdfclown.documents.interaction.actions.GoToEmbedded;
import org.pdfclown.documents.interaction.actions.GoToEmbedded.PathElement;
import org.pdfclown.documents.interaction.actions.GoToNonLocal;
import org.pdfclown.documents.interaction.actions.GoToURI;
import org.pdfclown.documents.interaction.navigation.document.Bookmark;
import org.pdfclown.documents.interaction.navigation.document.Bookmarks;
import org.pdfclown.documents.interaction.navigation.document.Destination;
import org.pdfclown.files.File;
import org.pdfclown.objects.PdfObjectWrapper;
import org.pdfclown.util.io.IOUtils;

/**
  This sample demonstrates <b>how to inspect the bookmarks</b> of a PDF document.

  @author Stefano Chizzolini (http://www.stefanochizzolini.it)
  @version 0.1.2.1, 04/08/15
*/
public class BookmarksParsingSample
  extends Sample
{
  @Override
  public void run(
    )
  {
    File file = null;
    try
    {
      // 1. Opening the PDF file...
      {
        String filePath = promptFileChoice("Please select a PDF file");
        try
        {file = new File(filePath);}
        catch(Exception e)
        {throw new RuntimeException(filePath + " file access error.",e);}
      }
      Document document = file.getDocument();

      // 2. Get the bookmarks collection!
      Bookmarks bookmarks = document.getBookmarks();
      if(!bookmarks.exists())
      {System.out.println("\nNo bookmark available (Outline dictionary not found).");}
      else
      {
        System.out.println("\nIterating through the bookmarks collection (please wait)...\n");
        // 3. Show the bookmarks!
        printBookmarks(bookmarks);
      }
    }
    finally
    {
      // 4. Closing the PDF file...
      IOUtils.closeQuietly(file);
    }
  }

  private void printBookmarks(
    Bookmarks bookmarks
    )
  {
    if(bookmarks == null)
      return;

    for(Bookmark bookmark : bookmarks)
    {
      // Show current bookmark!
      System.out.println("Bookmark '" + bookmark.getTitle() + "'");
      System.out.print("    Target: ");
      PdfObjectWrapper<?> target = bookmark.getTarget();
      if(target instanceof Destination)
      {printDestination((Destination)target);}
      else if(target instanceof Action)
      {printAction((Action)target);}
      else if(target == null)
      {System.out.println("[not available]");}
      else
      {System.out.println("[unknown type: " + target.getClass().getSimpleName() + "]");}

      // Show child bookmarks!
      printBookmarks(bookmark.getBookmarks());
    }
  }

  private void printAction(
    Action action
    )
  {
    System.out.println("Action [" + action.getClass().getSimpleName() + "] " + action.getBaseObject());
    if(action instanceof GoToDestination<?>)
    {
      if(action instanceof GoToNonLocal<?>)
      {
        FileSpecification<?> destinationFile = ((GoToNonLocal<?>)action).getDestinationFile();
        if(destinationFile != null)
        {System.out.println("      Filename: " + destinationFile.getPath());}

        if(action instanceof GoToEmbedded)
        {
          PathElement target = ((GoToEmbedded)action).getDestinationPath();
          System.out.println("      EmbeddedFilename: " + target.getEmbeddedFileName() + " Relation: " + target.getRelation());
        }
      }
      System.out.print("      ");
      printDestination(((GoToDestination<?>)action).getDestination());
    }
    else if(action instanceof GoToURI)
    {System.out.println("      URI: " + ((GoToURI)action).getURI());}
  }

  private void printDestination(
    Destination destination
    )
  {
    System.out.println(destination.getClass().getSimpleName() + " " + destination.getBaseObject());
    System.out.print("        Page ");
    Object pageRef = destination.getPage();
    if(pageRef instanceof Page)
    {
      Page page = (Page)pageRef;
      System.out.println(page.getNumber() + " [ID: " + page.getBaseObject() + "]");
    }
    else
    {System.out.println(((Integer)pageRef+1));}

    Object location = destination.getLocation();
    if(location != null)
    {System.out.println(String.format("        Location %s", location));}

    Double zoom = destination.getZoom();
    if(zoom != null)
    {System.out.println(String.format("        Zoom %s", zoom));}
  }
}